File: src/modules/reservation/reservationController.ts
/*
  File: reservationController.ts
  Author: Poon - Nunthatinn Veerapaiboon (nveerap@cmkl.ac.th)
  Description: Main reservation logics
    - Conflicts check
    - Recursively insert and edit reservations
  Lasted Modify: 2025-10-14 23.08
*/

import { randomUUID } from "crypto";
import { deleteReservationBySeriesId, getReservationsByTimeRange } from "./reservationService";
import { createReservationBySlotData } from "./reservationService";
import type { Slot } from "../types";

const WEEK_MS = 7 * 24 * 60 * 60 * 1000; // 1 week in milliseconds


// Helper function: check time overlap
function isOverlap(startA: Date, endA: Date, startB: Date, endB: Date): boolean {
    return startA < endB && endA > startB;
}


// Generate repeated weekly slots
// timeStart and timeEnd: timeslot of first week of reservation
// rep: number of weeks the reservation will repeat 
export function generateWeeklySlots(timeStart: Date, timeEnd: Date, rep: number): Slot[] {
  const slots: Slot[] = [];

  // Convert input strings to Date objects
  let currentStart = timeStart;
  let currentEnd = timeEnd;

  for (let i = 0; i < rep; i++) {
    slots.push({
      timeStart: currentStart,
      timeEnd: currentEnd,
    });

    // Increment date by 7 days, (Move to next week on same period)
    currentStart = new Date(currentStart.getTime() + WEEK_MS);
    currentEnd   = new Date(currentEnd.getTime()   + WEEK_MS);
  }

  return slots;
}


// Check conflicts on given time slots
async function checkConflicts(seriesId: string | null, roomId: string, slots: Slot[]): Promise<boolean> {
  if (!slots.length) return false;

  for (const slot of slots) {
    const slotStart = slot.timeStart;
    const dayStart = new Date(slotStart);
    const slotEnd = slot.timeEnd;
    dayStart.setHours(0, 0, 0, 0);
    const dayEnd   = new Date(slotStart);
    dayEnd.setHours(23, 59, 59, 999);
    try {
      const result = await getReservationsByTimeRange(roomId ,dayStart, dayEnd);

      if (!result.success) {
        console.error("Error checking slot:", slot, result.error);
      } else {
        for (const res of result.reservations) {
          // skip if checking itself series.
          if (seriesId && res.seriesId === seriesId) continue;
          
          const resStart = res.timeStart;
          const resEnd = res.timeEnd;
          // Check for overlap conflicts using helper function
          if (isOverlap(slotStart, slotEnd, resStart, resEnd)) {
            console.error(`Conflict detected: Room ${roomId} already booked from ${res.timeStart} to ${res.timeEnd}`);
            return true; // conflicts found
          }
        }
      }
    } catch (error) {
      console.error('Error in checkConflicts:', error);
    }
  }
  return false; // no conflicts
}



// ---------- Repeatation Reservation Logics ----------

// Add a new reservation series to the database.
export async function addReservation(roomId: string, timeStart: Date, timeEnd: Date, rep: number, competency: string): Promise<boolean> {
  const slots = generateWeeklySlots(timeStart, timeEnd, rep);
  const isConflict = await checkConflicts(null, roomId, slots);

  if (isConflict) {
    return false; // slot(s) overlap, reservation fail
  } else {
    const seriesId = randomUUID();
    const newSlotData = slots.map(slot => ({
      seriesId,
      roomId,
      timeStart: slot.timeStart,
      timeEnd: slot.timeEnd,
      competency,
    }));
    try {
      // Insert slots into database
      await createReservationBySlotData(newSlotData);
      console.log("Reservations created successfully");
      return true;
    } catch (error) {
      console.error('Error inserting reservations:', error);
      return false;
    }
  }
}


// Edit an existing reservation series by seriesId.
export async function editReservation(seriesId: string, roomId: string, timeStart: Date, timeEnd: Date, rep: number, competency: string): Promise<boolean> {
  const slots = generateWeeklySlots(timeStart, timeEnd, rep);
  const isConflict = await checkConflicts(seriesId, roomId, slots);

  if (isConflict) {
    return false; // slot overlap, reservation fail
  } else {
    // Insert slots into database
    const newSlotData = slots.map(slot => ({
      seriesId,
      roomId,
      timeStart: slot.timeStart,
      timeEnd: slot.timeEnd,
      competency,
    }));
    try {
      // delete existing reservation series
      await deleteReservationBySeriesId(seriesId);
      // add new reservation series with same seriesId
      await createReservationBySlotData(newSlotData);

      console.log("Reservations edited successfully");
      return true;
    } catch (error) {
      console.error('Error editing reservations:', error);
      return false;
    }
  }
}


// Delete all reservations in a series by seriesId.
export async function deleteReservation(seriesId: string): Promise<boolean> {
  try {
    await deleteReservationBySeriesId(seriesId);
    console.log("Reservations deleted successfully");
    return true;
  } catch (error) {
    console.error('Error inserting reservations:', error);
      return false;
  }
}

File: classroom_reservation/lib/services/reservation_controller.dart
import 'package:uuid/uuid.dart';
import 'database_services/reservation_services.dart';
import 'database_services/series_services.dart';
import '../models/reservation.dart';

const int WEEK_MS = 7 * 24 * 60 * 60 * 1000;

class Slot {
  DateTime timeStart;
  DateTime timeEnd;

  Slot({required this.timeStart, required this.timeEnd});
}

// Generate repeated weekly slots
List<Slot> generateWeeklySlots(DateTime timeStart, DateTime timeEnd, int rep) {
  List<Slot> slots = [];

  DateTime currentStart = timeStart;
  DateTime currentEnd = timeEnd;

  for (int i = 0; i < rep; i++) {
    slots.add(Slot(timeStart: currentStart, timeEnd: currentEnd));
    currentStart = currentStart.add(Duration(milliseconds: WEEK_MS));
    currentEnd = currentEnd.add(Duration(milliseconds: WEEK_MS));
  }

  return slots;
}

// Check overlap
bool isOverlap(DateTime startA, DateTime endA, DateTime startB, DateTime endB) {
  return startA.isBefore(endB) && endA.isAfter(startB);
}


Future<bool> checkConflicts(
  ReservationServices reservationServices,
  String? seriesId,
  String roomId,
  List<Slot> slots,
) async {
  if (slots.isEmpty) return false;

  for (var slot in slots) {
    DateTime slotStart = slot.timeStart;
    DateTime slotEnd = slot.timeEnd;

    DateTime dayStart = DateTime(slotStart.year, slotStart.month, slotStart.day, 0, 0, 0);
    DateTime dayEnd   = DateTime(slotStart.year, slotStart.month, slotStart.day, 23, 59, 59);

    var result = await reservationServices.getReservationsByTimeRange(
      roomId: roomId,
      startTime: dayStart,
      endTime: dayEnd,
    );

    if (!result['success']) {
      print('Error checking slot: $slot, ${result['error']}');
    } else {
      List<Reservation> reservations = result['reservations'];
      for (var res in reservations) {
        if (seriesId != null && res.seriesId == seriesId) continue;
        if (isOverlap(slotStart, slotEnd, res.timeStart, res.timeEnd)) {
          print('Conflict detected: Room $roomId already booked from ${res.timeStart} to ${res.timeEnd}');
          return true;
        }
      }
    }
  }
  return false;
}


Future<bool> addReservation(
  ReservationServices reservationServices,
  SeriesServices seriesServices,
  String roomId,
  DateTime timeStart,
  DateTime timeEnd,
  int capacity,
  int rep,
  String competency,
) async {
  List<Slot> slots = generateWeeklySlots(timeStart, timeEnd, rep);
  bool isConflict = await checkConflicts(reservationServices, null, roomId, slots);

  if (isConflict) return false;

  String seriesId = Uuid().v4(); // generate unique ID
  List<Reservation> newReservations = slots.map((slot) {
    return Reservation(
      seriesId: seriesId,
      roomId: roomId,
      timeStart: slot.timeStart,
      timeEnd: slot.timeEnd,
      competency: competency,
    );
  }).toList();

  try {
    await reservationServices.createReservations(newReservations);
    await seriesServices.createSeries(seriesId, capacity, rep);
    print('Reservations created successfully');
    return true;
  } catch (error) {
    print('Error inserting reservations: $error');
    return false;
  }
}


Future<bool> editReservation(
  ReservationServices reservationServices,
  SeriesServices seriesServices,
  String seriesId,
  String roomId,
  DateTime timeStart,
  DateTime timeEnd,
  int capacity,
  int rep,
  String competency,
) async {
  List<Slot> slots = generateWeeklySlots(timeStart, timeEnd, rep);
  bool isConflict = await checkConflicts(reservationServices, seriesId, roomId, slots);

  if (isConflict) return false;

  List<Reservation> newReservations = slots.map((slot) {
    return Reservation(
      seriesId: seriesId,
      roomId: roomId,
      timeStart: slot.timeStart,
      timeEnd: slot.timeEnd,
      competency: competency,
    );
  }).toList();

  try {
    await reservationServices.deleteReservationsBySeriesId(seriesId);
    await reservationServices.createReservations(newReservations);
    await seriesServices.editSeriesById(seriesId, capacity, rep);
    print('Reservations edited successfully');
    return true;
  } catch (error) {
    print('Error editing reservations: $error');
    return false;
  }
}


Future<bool> deleteReservation(
  ReservationServices reservationServices,
  SeriesServices seriesServices,
  String seriesId,
) async {
  try {
    await reservationServices.deleteReservationsBySeriesId(seriesId);
    await seriesServices.deleteSeriesById(seriesId);
    print('Reservations deleted successfully');
    return true;
  } catch (error) {
    print('Error deleting reservations: $error');
    return false;
  }
}

